src/java.base/share/classes/java/net/DatagramSocket.java

```diff
@@ -1038,6 +1224,95 @@ public DatagramChannel getChannel() {
         return delegate().supportedOptions();
     }
 
+    /**
+     * Joins the specified multicast group at the specified interface.
+     *
+     * <p> In order to join a multicast group, the caller should specify
+     * the IP address of the multicast group to join, and the local
+     * {@linkplain NetworkInterface network interface} to receive multicast
+     * packets from.
+     * <ul>
+     *  <li> The {@code mcastaddr} argument indicates the IP address
+     *   of the multicast group to join. For historical reasons this is
+     *   specified as a {@code SocketAddress}.
+     *   The default implementation only supports {@link InetSocketAddress} and
+     *   the {@link InetSocketAddress#getPort() port} information is ignored.
+     *  </li>
+     *  <li> The {@code netIf} argument specifies the local interface to receive
+     *       multicast datagram packets, or {@code null} to defer to the interface
+     *       set for outgoing multicast datagrams.
+     *       If {@code null}, and no interface has been set, the behaviour is
+     *       unspecified: any interface may be selected or the operation may fail
+     *       with a {@code SocketException}.
+     *  </li>
+     * </ul>
+     *
+     * <p>If there is a security manager, this method first
+     * calls its {@code checkMulticast} method with the {@code mcastaddr}
+     * argument as its argument.
+     *
+     * @apiNote The default interface for sending outgoing multicast datagrams
+     * can be configured with {@link #setOption(SocketOption, Object)}
+     * with {@link StandardSocketOptions#IP_MULTICAST_IF}.
+     *
+     * @param  mcastaddr indicates the multicast address to join.
+     * @param  netIf specifies the local interface to receive multicast
+     *         datagram packets, or {@code null}.
+     * @throws IOException if there is an error joining, or when the address
+     *         is not a multicast address, or the platform does not support
+     *         multicasting
+     * @throws SecurityException if a security manager exists and its
+     *         {@code checkMulticast} method doesn't allow the join.
+     * @throws IllegalArgumentException if mcastaddr is {@code null} or is a
+     *         SocketAddress subclass not supported by this socket
+     * @see    SecurityManager#checkMulticast(InetAddress)
+     * @see    DatagramChannel#join(InetAddress, NetworkInterface)
+     * @see    StandardSocketOptions#IP_MULTICAST_IF
+     * @since  17
+     */
+    public void joinGroup(SocketAddress mcastaddr, NetworkInterface netIf)
```

Comment:
What would you think about reducing the first sentence to "Join a multicast group".  The reason is "at the specified interface" is a bit awkward in that it identifies the network by its network interface and that is hard to word in a readable-way. The second paragraph where it explains the two parameters is good so that helps the case for a simple first sentence.

I think we need wording to make it clear joinGroup can be call several times to join several multicast groups. It can join the same group on several networks too.

Another thing that is looks to be missing is the behavior when already a member of the group. We might have to tweak the wording for leaveGroup to make it clear how it behaves when invoked to leave a group that the socket didn't join.